Source/Core/DolphinWX/LuaScripting.h

```diff
@@ -0,0 +1,81 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <map>
+#include <mutex>
+#include <wx/frame.h>
+
+// Forward Declarations
+class wxString;
+class wxFileDialog;
+class wxTextCtrl;
+class wxStaticText;
+class wxMenuItem;
+class wxMenuBar;
+class wxMenu;
+class wxButton;
+struct GCPadStatus;
+struct lua_State;
+
+namespace Lua
+{
+using LuaFunction = int (*)(lua_State* L);
+
+class LuaScriptFrame;
+
+extern std::map<const char*, LuaFunction> m_registered_functions;
+
+class LuaThread final : public wxThread
+{
+public:
+  LuaThread(LuaScriptFrame* p, const wxString& file);
+  ~LuaThread();
+  static GCPadStatus* GetPadStatus();
+  static std::mutex* GetMutex();
+
+private:
+  LuaScriptFrame* m_parent = nullptr;
+  wxString m_file_path;
+  wxThread::ExitCode Entry() override;
+  static GCPadStatus m_pad_status;

```

Comment:
How is this any different from having global state behind a namespace?

`LuaThread::GetMutex()` is essentially just acting as a qualifier that needs to be used instead of directly accessing the variable.

There should be no static mutable state on a thread or anything else related to the implementation. static state like this is very indicative of a code smell or design issue that doesn't properly have it's operations and access points fully sorted out.

It doesn't really make sense from a behavioral standpoint to be able to set or modify information related to a thread in the absence of an actual instance of said thread.

Reply:
The simple fact is that LuaScriptFrame and LuaThread must both be able to lock and unlock a shared mutex so that script execution cannot halt in the middle of a call to LuaScriptFrame::GetValues().

I understand the problem with the code, and I don't really know where this comment is going. I'm just frustrated.